src/opuntia.erl

```diff
@@ -1,53 +1,106 @@
+%% This module implements the token-bucket traffic-shaping algorithm.
+%%
+%% The rate is given in tokens per time unit and a bucket size.
+%%
+%% The delay is always returned in milliseconds unit,
+%% as this is the unit receives and timers use in the BEAM.
 %% @end
 -module(opuntia).
 
 -export([new/1, update/2]).
 
 -include("opuntia.hrl").
 
--type timestamp() :: number().
--type tokens() :: non_neg_integer().
 -type delay() :: non_neg_integer().
+%% Number of milliseconds that is advise to wait after a shaping update.
+
 -type rate() :: non_neg_integer().
--type shaper() :: #token_bucket{} | none.
+%% Number of tokens accepted per `time_unit'.
+
+-type time_unit() :: second | millisecond | microsecond.
+%% Supported shaping time units: `second', `millisecond', `microsecond'.
+
+-type bucket_size() :: non_neg_integer().
+%% Maximum capacity of the bucket regardless of how much time it passes.
+
+-type tokens() :: non_neg_integer().
+%% Unit element the shaper consumes, for example bytes or requests.
+
+-type shape() :: 0 | {bucket_size(), rate(), time_unit()}.
+%% Zero, or a rate-per-time-unit and a maximum number of tokens. See `new/1' for more details.
+
+-type shaper() :: none | #token_bucket_shaper{}.
+%% Shaper type
+
+-export_type([shaper/0, shape/0, tokens/0, time_unit/0, delay/0]).
 
--export_type([tokens/0, timestamp/0, delay/0, rate/0, shaper/0]).
+-define(NON_NEG_INT(N), (is_integer(N) andalso N > 0)).
+-define(TU(T), (second =:= T orelse millisecond =:= T orelse microsecond =:= T)).
 
--spec new(rate()) -> shaper().
+%% @doc Creates a new shaper according to the configuration.
+%%
+%% If zero is given, no shaper in created and any update action will always return zero delay;
+%% to configure a shaper it will need a `{MaximumTokens, Rate, TimeUnit}', where
+%% <ul>
+%% <li>`TimeUnit' is the time unit of measurement, that is, `second', `millisecond', or `microsecond'.</li>
+%% <li>`Rate' is the number of tokens per `TimeUnit' the bucket will grow with.</li>
+%% <li>`MaximumTokens' is the maximum number of tokens the bucket can grow.</li>
+%% </ul>
+%%
+%% So, for example, if we configure `{60000, 10, millisecond}', it means that the bucket will
+%% allow `10' tokens per `millisecond', up to 60000 tokens so regardless of how long it is left
+%% unused to charge, it will never charge further than 60000 tokens.
+%%
+%% Note that shapers start unloaded, that is, without available tokens.
+-spec new(shape()) -> shaper().
 new(0) ->
     none;
-new(MaxRatePerMs) ->
-    #token_bucket{rate = MaxRatePerMs,
-                  available_tokens = MaxRatePerMs,
-                  last_update = erlang:monotonic_time(millisecond)}.
+new({MaximumTokens, Rate, TimeUnit})
+  when ?NON_NEG_INT(MaximumTokens), ?NON_NEG_INT(Rate), ?TU(TimeUnit) ->
+    #token_bucket_shaper{shape = {MaximumTokens, Rate, TimeUnit},
+                         available_tokens = 0,
```

Comment:
This rings an alarm bell instantly, so I tested it:

```
23> f(Sh),Sh=opuntia:new({100000,10000,second}),opuntia:update(Sh,0).
{{token_bucket_shaper,{100000,10000,second},0,-576460334},0}
24> f(Sh),Sh=opuntia:new({100000,10000,second}),opuntia:update(Sh,1).
{{token_bucket_shaper,{100000,10000,second},0,-576460099},1}
```

And I was right - the delay kicks in instantly, and you need a second to unblock.
So to continue my review I needed to work around it:

```
available_tokens = Rate,
```

Reply:
The idea is that this wasn't necessarily wrong but whether you want the shapers to start charged or uncharged, it's a matter of deciding the API. Easy to change, only tests need to be adjusted.

Reply:
Charging them was just a workaround. the main issue here is that if unit is `second`, then for up to a second I cannot use any tokens even though e.g. after 500 ms I should have the bucket halfway charged. So the unit starts acting like the counter resolution, which IMO shouldn't be the case.

Reply:
Hmm... good point, but, that's more or less the contract too and why I separated bucket size from the rate. In a way that's what we had implicitly in MongooseIM, we had for example 60k tokens (bytes) per second but recharging was happening in milliseconds resolution, up to a maximum of 60k tokens regardless of how much time it had passed.

Also, the delay should always be given in milliseconds, not in the resolution you requested, all BEAM timers work in milliseconds so that's why I chose that.